### PR TITLE
Interface reference docs

### PR DESCRIPTION
**Pull request by akrmn**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/daml/reference/interfaces.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+.. literalinclude:: ../code-snippets-dev/Interfaces.daml
+   :language: daml
+   :start-after: -- INTERFACE_METHODS_BEGIN
+   :end-before: -- INTERFACE_METHODS_END
+
+- An interface may define any number of methods
+- Methods are visible as functions at the top level and in interface choices
+  (see :ref:`interface-choices` below).
+- The type of a method will always be prefixed with
```
Comment (akrmn) [inline docs/source/daml/reference/interfaces.rst:42]: there must be a less convoluted way to say this but I can't think of one right now

File: docs/source/daml/reference/interfaces.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
+.. Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+.. _interfaces:
+
+Reference: Interfaces
+#####################
+
+.. warning::
+  This feature is under active development and not officially supported in
```
Comment (akrmn) [inline docs/source/daml/reference/interfaces.rst:10]: Do we have a standard phrasing for things like this?

File: docs/source/daml/reference/interfaces.rst [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+Interface precondition
+----------------------
+
+.. literalinclude:: ../code-snippets-dev/Interfaces.daml
+   :language: daml
+   :start-after: -- INTERFACE_ENSURE_BEGIN
+   :end-before: -- INTERFACE_ENSURE_END
+
+- Introduced with the keyword ``ensure``, must be a boolean expression.
```
Comment (robin-da) [inline docs/source/daml/reference/interfaces.rst:64]: I'd change this into a whole sentence, like the other bullet points.

File: docs/source/daml/reference/interfaces.rst [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+- It is possible to define interfaces without an ``ensure`` clause, but writing
+  more than one is a compilation error.
+- ``this`` is in scope in the method with the type of the interface.
+  ``self``, however, is not.
+- The interface methods can be used as part of the expression (e.g. ``method1``).
+- It is evaluated and checked together with the implementing template's
```
Comment (robin-da) [inline docs/source/daml/reference/interfaces.rst:70]: It is evaluated and checked right after the implementing template precondition.

File: docs/source/daml/reference/interfaces.rst [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+Interface choices
+-----------------
+
+.. literalinclude:: ../code-snippets-dev/Interfaces.daml
+   :language: daml
+   :start-after: -- INTERFACE_CHOICE_BEGIN
+   :end-before: -- INTERFACE_CHOICE_END
+
+- Works in a very similar way to template choices. Any contract of an
```
Comment (robin-da) [inline docs/source/daml/reference/interfaces.rst:83]: Interface choices work in a very similar way to ...

File: docs/source/daml/reference/interfaces.rst [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+   :language: daml
+   :start-after: -- INTERFACE_REQUIRES_BEGIN
+   :end-before: -- INTERFACE_REQUIRES_END
+
+- An interface can depend on other interfaces. These are specified with the
+  ``requires`` keyword after the interface name but before the
+  ``where`` keyword, separated by commas.
+- For a template's implementation of an interface to be valid, all its required
+  interfaces must also be implemented by the template.
+- If the interface doesn't have any methods, precondition or choices,
```
Comment (robin-da) [inline docs/source/daml/reference/interfaces.rst:119]: This was stated in the previous section already.

**Review by robin-da — APPROVED** [review]

**Pull request merged**